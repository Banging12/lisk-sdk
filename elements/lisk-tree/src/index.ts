/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { verifyProof } from './merkle_tree/verify_proof';
import { calculateRootFromUpdateData } from './merkle_tree/calculate';
import { MerkleTree } from './merkle_tree/merkle_tree';
import { calculateMerkleRoot, calculateMerkleRootWithLeaves } from './merkle_tree/utils';

export const regularMerkleTree = {
	verifyProof,
	calculateRootFromUpdateData,
	calculateMerkleRoot,
	calculateMerkleRootWithLeaves,
	MerkleTree,
};

export { MerkleTree };
