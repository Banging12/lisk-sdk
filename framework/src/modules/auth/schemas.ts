/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const authAccountSchema = {
	type: 'object',
	properties: {
		nonce: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		numberOfSignatures: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		mandatoryKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 3,
		},
		optionalKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 4,
		},
	},
	required: ['nonce', 'numberOfSignatures', 'mandatoryKeys', 'optionalKeys'],
};

export const registerMultisignatureCommandSchema = {
	$id: '/auth/command/regMultisig',
	type: 'object',
	properties: {
		numberOfSignatures: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
		mandatoryKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 2,
		},
		optionalKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 3,
		},
	},
	required: ['numberOfSignatures', 'mandatoryKeys', 'optionalKeys'],
};

export const configSchema = {
	$id: '/auth/config',
	type: 'object',
	properties: {},
};
