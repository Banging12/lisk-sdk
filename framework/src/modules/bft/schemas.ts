/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const bftModuleConfig = {
	$id: 'modules/bft/config',
	type: 'object',
	required: ['batchSize'],
	properties: {
		batchSize: {
			type: 'integer',
			format: 'uint32',
		},
	},
};

export interface BFTParametersValidator {
	address: Buffer;
	bftWeight: bigint;
}

export interface BFTParameters {
	prevoteThreshold: bigint;
	precommitThreshold: bigint;
	certificateThreshold: bigint;
	validators: BFTParametersValidator[];
	validatorsHash: Buffer;
}

export const bftParametersSchema = {
	$id: 'modules/bft/bftParameters',
	type: 'object',
	required: [
		'prevoteThreshold',
		'precommitThreshold',
		'certificateThreshold',
		'validators',
		'validatorsHash',
	],
	properties: {
		prevoteThreshold: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		precommitThreshold: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		certificateThreshold: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		validators: {
			type: 'array',
			fieldNumber: 4,
			items: {
				type: 'object',
				required: ['address', 'bftWeight'],
				properties: {
					address: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					bftWeight: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
		},
		validatorsHash: {
			dataType: 'bytes',
			fieldNumber: 5,
		},
	},
};

export interface BFTVotesBlockInfo {
	height: number;
	generatorAddress: Buffer;
	maxHeightGenerated: number;
	maxHeightPrevoted: number;
	prevoteWeight: number;
	precommitWeight: number;
}

export interface BFTVotesActiveValidatorsVoteInfo {
	address: Buffer;
	minActiveHeight: number;
	largestHeightPrecommit: number;
}

export interface BFTVotes {
	maxHeightPrevoted: number;
	maxHeightPrecommitted: number;
	maxHeightCertified: number;
	blockBFTInfos: BFTVotesBlockInfo[];
}

export const bftVotesSchema = {
	$id: 'modules/bft/bftVotes',
	type: 'object',
	required: [
		'maxHeightPrevoted',
		'maxHeightPrecommitted',
		'maxHeightCertified',
		'blockBFTInfos',
		'activeValidatorsVoteInfo',
	],
	properties: {
		maxHeightPrevoted: { dataType: 'uint32', fieldNumber: 1 },
		maxHeightPrecommitted: { dataType: 'uint32', fieldNumber: 2 },
		maxHeightCertified: { dataType: 'uint32', fieldNumber: 3 },
		blockBFTInfos: {
			type: 'array',
			fieldNumber: 4,
			items: {
				type: 'object',
				required: [
					'height',
					'generatorAddress',
					'maxHeightGenerated',
					'maxHeightPrevoted',
					'prevoteWeight',
					'precommitWeight',
				],
				properties: {
					height: { dataType: 'uint32', fieldNumber: 1 },
					generatorAddress: { dataType: 'bytes', fieldNumber: 2 },
					maxHeightGenerated: { dataType: 'uint32', fieldNumber: 3 },
					maxHeightPrevoted: { dataType: 'uint32', fieldNumber: 4 },
					prevoteWeight: { dataType: 'uint64', fieldNumber: 5 },
					precommitWeight: { dataType: 'uint64', fieldNumber: 6 },
				},
			},
		},
		activeValidatorsVoteInfo: {
			type: 'array',
			fieldNumber: 5,
			items: {
				type: 'object',
				required: ['address', 'minActiveHeight', 'largestHeightPrecommit'],
				properties: {
					address: { dataType: 'bytes', fieldNumber: 1 },
					minActiveHeight: { dataType: 'uint32', fieldNumber: 2 },
					largestHeightPrecommit: { dataType: 'uint32', fieldNumber: 3 },
				},
			},
		},
	},
};
