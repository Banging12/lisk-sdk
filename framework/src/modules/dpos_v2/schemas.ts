/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const voterStoreSchema = {
	type: 'object',
	required: ['sentVotes', 'pendingUnlocks'],
	properties: {
		sentVotes: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['delegateAddress', 'amount'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
		},
		pendingUnlocks: {
			type: 'array',
			fieldNumber: 2,
			items: {
				type: 'object',
				required: ['delegateAddress', 'amount', 'unvoteHeight'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
					unvoteHeight: {
						dataType: 'uint32',
						fieldNumber: 3,
					},
				},
			},
		},
	},
};

export const delegateStoreSchema = {
	type: 'object',
	required: [
		'name',
		'totalVotesReceived',
		'selfVotes',
		'lastGeneratedHeight',
		'isBanned',
		'pomHeights',
		'consecutiveMissedBlocks',
	],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
		},
		totalVotesReceived: {
			dataType: 'uint64',
			fieldNumber: 2,
		},
		selfVotes: {
			dataType: 'uint64',
			fieldNumber: 3,
		},
		lastGeneratedHeight: {
			dataType: 'uint32',
			fieldNumber: 4,
		},
		isBanned: {
			dataType: 'boolean',
			fieldNumber: 5,
		},
		pomHeights: {
			type: 'array',
			fieldNumber: 6,
			items: { dataType: 'uint32' },
		},
		consecutiveMissedBlocks: {
			dataType: 'uint32',
			fieldNumber: 7,
		},
	},
};

export const nameStoreSchema = {
	type: 'object',
	required: ['delegateAddress'],
	properties: {
		delegateAddress: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
};

export const snapshotStoreSchema = {
	type: 'object',
	required: ['activeDelegates', 'delegateWeightSnapshot'],
	properties: {
		activeDelegates: {
			type: 'array',
			fieldNumber: 1,
			items: { dataType: 'bytes' },
		},
		delegateWeightSnapshot: {
			type: 'array',
			fieldNumber: 2,
			items: {
				type: 'object',
				required: ['delegateAddress', 'delegateWeight'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					delegateWeight: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
				},
			},
		},
	},
};

export const genesisDataStoreSchema = {
	type: 'object',
	required: ['height', 'initRounds', 'initDelegates'],
	properties: {
		height: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		initRounds: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		initDelegates: {
			type: 'array',
			fieldNumber: 3,
			items: { dataType: 'bytes' },
		},
	},
};

export const previousTimestampStoreSchema = {
	type: 'object',
	required: ['timestamp'],
	properties: {
		timestamp: {
			dataType: 'uint32',
			fieldNumber: 1,
		},
	},
};

export const delegateRegistrationCommandParamsSchema = {
	$id: '/dpos/command/registerDelegateParams',
	type: 'object',
	required: ['name', 'generatorKey', 'blsKey', 'proofOfPossession'],
	properties: {
		name: {
			dataType: 'string',
			fieldNumber: 1,
		},
		generatorKey: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
		blsKey: {
			dataType: 'bytes',
			fieldNumber: 3,
		},
		proofOfPossession: {
			dataType: 'bytes',
			fieldNumber: 4,
		},
	},
};

export const updateGeneratorKeyCommandParamsSchema = {
	$id: '/dpos/command/updateGeneratorKeyParams',
	type: 'object',
	required: ['generatorKey'],
	properties: {
		generatorKey: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
	},
};

export const voteCommandParamsSchema = {
	$id: '/dpos/command/voteDelegateParams',
	type: 'object',
	required: ['votes'],
	properties: {
		votes: {
			type: 'array',
			fieldNumber: 1,
			items: {
				type: 'object',
				required: ['delegateAddress', 'amount'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
					},
					amount: {
						dataType: 'sint64',
						fieldNumber: 2,
					},
				},
			},
		},
	},
};

export const pomCommandParamsSchema = {
	$id: '/dpos/command/reportDelegateMisbehaviorParams',
	type: 'object',
	required: ['header1', 'header2'],
	properties: {
		header1: {
			dataType: 'bytes',
			fieldNumber: 1,
		},
		header2: {
			dataType: 'bytes',
			fieldNumber: 2,
		},
	},
};

export const unlockCommandParamsSchema = {
	$id: '/dpos/command/unlockTokenParams',
	type: 'object',
	properties: {},
};

export const configSchema = {
	$id: '/dpos/config',
	type: 'object',
	properties: {
		factorSelfVotes: {
			dataType: 'uint32',
		},
		maxLengthName: {
			dataType: 'uint32',
		},
		maxNumberSentVotes: {
			dataType: 'uint32',
		},
		maxNumberPendingUnlocks: {
			dataType: 'uint32',
		},
		failSafeMissedBlocks: {
			dataType: 'uint32',
		},
		failSafeInactiveWindow: {
			dataType: 'uint32',
		},
		punishmentWindow: {
			dataType: 'uint32',
		},
		roundLength: {
			dataType: 'uint32',
		},
		bftThreshold: {
			dataType: 'uint32',
		},
		minWeightStandby: {
			dataType: 'uint32',
		},
		numberActiveDelegates: {
			dataType: 'uint32',
		},
		tokenIDDPoS: {
			type: 'object',
			properties: {
				chainID: {
					dataType: 'uint32',
				},
				localID: {
					dataType: 'uint32',
				},
			},
			required: ['chainID', 'localID'],
		},
	},
	required: [
		'factorSelfVotes',
		'maxLengthName',
		'maxNumberSentVotes',
		'maxNumberPendingUnlocks',
		'failSafeMissedBlocks',
		'failSafeInactiveWindow',
		'punishmentWindow',
		'roundLength',
		'bftThreshold',
		'minWeightStandby',
		'numberActiveDelegates',
		'tokenIDDPoS',
	],
};
