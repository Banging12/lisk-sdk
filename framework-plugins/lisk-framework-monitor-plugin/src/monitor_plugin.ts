/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Server } from 'http';
import { Block } from '@liskhq/lisk-chain';
import { validator } from '@liskhq/lisk-validator';
import { BaseChannel, BasePlugin } from 'lisk-framework';
import * as express from 'express';
import type { Express } from 'express';
import * as cors from 'cors';
import * as rateLimit from 'express-rate-limit';
import * as middlewares from './middlewares';
import { MonitorPluginConfig, SharedState } from './types';
import * as controllers from './controllers';
import { transactionAnnouncementSchema, postBlockEventSchema, configSchema } from './schemas';
import { Endpoint } from './endpoint';

interface BlockData {
	readonly block: string;
}

export class MonitorPlugin extends BasePlugin<MonitorPluginConfig> {
	public name = 'monitor';
	public configSchema = configSchema;
	public endpoint = new Endpoint();

	private _server!: Server;
	private _app!: Express;
	private _state!: SharedState;

	public get nodeModulePath(): string {
		return __filename;
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async load(_channel: BaseChannel): Promise<void> {
		this._app = express();

		this._state = {
			forks: {
				forkEventCount: 0,
				blockHeaders: {},
			},
			transactions: {},
			blocks: {},
		};

		this._registerMiddlewares(this.config);
		this._registerControllers();
		this._registerAfterMiddlewares(this.config);
		this._subscribeToEvents();
		this._server = this._app.listen(this.config.port, this.config.host);
	}

	public async unload(): Promise<void> {
		await new Promise<void>((resolve, reject) => {
			this._server.close(err => {
				if (err) {
					reject(err);
					return;
				}
				resolve();
			});
		});
	}

	public get state(): SharedState {
		return this._state;
	}

	private _registerMiddlewares(options: MonitorPluginConfig): void {
		// Register middlewares
		this._app.use(cors(options.cors));
		this._app.use(express.json());
		this._app.use(rateLimit(options.limits));
		this._app.use(middlewares.whiteListMiddleware(options));
	}

	private _registerAfterMiddlewares(_options: MonitorPluginConfig): void {
		this._app.use(middlewares.errorMiddleware());
	}

	private _registerControllers(): void {
		this._app.get(
			'/api/prometheus/metrics',
			controllers.prometheusExport.getData(this.apiClient, this._state),
		);
	}

	private _subscribeToEvents(): void {
		this.apiClient.subscribe('app_networkEvent', (eventData?: Record<string, unknown>) => {
			const { event, data } = eventData as { event: string; data: unknown };

			if (event === 'postTransactionsAnnouncement') {
				const errors = validator.validate(
					transactionAnnouncementSchema,
					data as Record<string, unknown>,
				);
				if (errors.length > 0) {
					return;
				}
				this._handlePostTransactionAnnounce(data as { transactionIds: string[] });
			}

			if (event === 'postBlock') {
				const errors = validator.validate(postBlockEventSchema, data as Record<string, unknown>);
				if (errors.length > 0) {
					return;
				}
				this._handlePostBlock(data as BlockData);
			}
		});

		this.apiClient.subscribe('app_chainForked', (data?: Record<string, unknown>) => {
			const { block } = (data as unknown) as BlockData;
			this._handleFork(block);
		});
	}

	private _handlePostTransactionAnnounce(data: { transactionIds: string[] }) {
		for (const aTransactionId of data.transactionIds) {
			if (this._state.transactions[aTransactionId]) {
				this._state.transactions[aTransactionId].count += 1;
			} else {
				this._state.transactions[aTransactionId] = {
					count: 1,
					timeReceived: Date.now(),
				};
				this._cleanUpTransactionStats();
			}
		}
	}

	private _cleanUpTransactionStats() {
		const expiryTime = 600000;
		for (const transactionID of Object.keys(this._state.transactions)) {
			if (Date.now() - this._state.transactions[transactionID].timeReceived > expiryTime) {
				delete this._state.transactions[transactionID];
			}
		}
	}

	private _handleFork(block: string) {
		this._state.forks.forkEventCount += 1;
		const { header } = Block.fromBytes(Buffer.from(block, 'hex'));
		const blockId = header.id.toString('hex');
		if (this._state.forks.blockHeaders[blockId]) {
			this._state.forks.blockHeaders[blockId].timeReceived = Date.now();
		} else {
			this._state.forks.blockHeaders[blockId] = {
				blockHeader: header,
				timeReceived: Date.now(),
			};
		}
	}

	private _handlePostBlock(data: BlockData) {
		const decodedBlock = Block.fromBytes(Buffer.from(data.block, 'hex'));

		if (!this._state.blocks[decodedBlock.header.id.toString('hex')]) {
			this._state.blocks[decodedBlock.header.id.toString('hex')] = {
				count: 0,
				height: decodedBlock.header.height,
			};
		}

		this._state.blocks[decodedBlock.header.id.toString('hex')].count += 1;

		// Clean up blocks older than current height minus 300 blocks
		for (const id of Object.keys(this._state.blocks)) {
			const blockInfo = this._state.blocks[id];
			if (blockInfo.height < decodedBlock.header.height - 300) {
				delete this._state.blocks[id];
			}
		}
	}
}
