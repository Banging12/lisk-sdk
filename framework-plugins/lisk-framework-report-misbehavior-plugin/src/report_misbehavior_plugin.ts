/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import {
	BasePlugin,
	BaseChannel,
	PluginInitContext,
	db as liskDB,
	codec,
	validator as liskValidator,
	chain,
	cryptography,
} from 'lisk-sdk';
import {
	getDBInstance,
	saveBlockHeaders,
	getContradictingBlockHeader,
	clearBlockHeaders,
} from './db';
import { ReportMisbehaviorPluginConfig, State } from './types';
import { postBlockEventSchema, configSchema } from './schemas';
import { Endpoint } from './endpoint';

const { getAddressAndPublicKeyFromPassphrase, getAddressFromPassphrase, signData } = cryptography;
const { BlockHeader, Transaction, TAG_TRANSACTION } = chain;
const { validator } = liskValidator;

export class ReportMisbehaviorPlugin extends BasePlugin<ReportMisbehaviorPluginConfig> {
	public name = 'reportMisbehavior';
	public configSchema = configSchema;
	public endpoint = new Endpoint();

	private _pluginDB!: liskDB.KVStore;
	private readonly _state: State = { currentHeight: 0 };
	private _clearBlockHeadersIntervalId!: NodeJS.Timer | undefined;

	public get nodeModulePath(): string {
		return __filename;
	}

	public async init(context: PluginInitContext): Promise<void> {
		await super.init(context);
		this.endpoint.init(this._state, this.config);
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async load(_channel: BaseChannel): Promise<void> {
		// TODO: https://github.com/LiskHQ/lisk-sdk/issues/6201
		this._pluginDB = await getDBInstance(this.dataPath);
		// Listen to new block and delete block events
		this._subscribeToChannel();
		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._clearBlockHeadersIntervalId = setInterval(() => {
			clearBlockHeaders(this._pluginDB, this._state.currentHeight).catch(error =>
				this.logger.error(error),
			);
		}, this.config.clearBlockHeadersInterval);
	}

	public async unload(): Promise<void> {
		clearInterval(this._clearBlockHeadersIntervalId as NodeJS.Timer);

		await this._pluginDB.close();
	}

	private _subscribeToChannel(): void {
		this.apiClient.subscribe('app_networkEvent', async (eventData?: Record<string, unknown>) => {
			const { event, data } = eventData as { event: string; data: unknown };

			if (event === 'postBlock') {
				const errors = validator.validate(postBlockEventSchema, data as Record<string, unknown>);
				if (errors.length > 0) {
					this.logger.error(errors, 'Invalid block data');
					return;
				}
				const blockData = data as { block: string };
				const { header } = codec.decode<chain.RawBlock>(
					this.apiClient.schemas.block,
					Buffer.from(blockData.block, 'hex'),
				);
				try {
					const saved = await saveBlockHeaders(this._pluginDB, header);
					if (!saved) {
						return;
					}
					const decodedBlockHeader = BlockHeader.fromBytes(header);

					// Set new currentHeight
					if (decodedBlockHeader.height > this._state.currentHeight) {
						this._state.currentHeight = decodedBlockHeader.height;
					}
					const contradictingBlock = await getContradictingBlockHeader(
						this._pluginDB,
						decodedBlockHeader,
						this.apiClient,
					);
					if (contradictingBlock && this._state.passphrase) {
						const encodedTransaction = await this._createPoMTransaction(
							decodedBlockHeader,
							contradictingBlock,
						);
						const result = await this.apiClient.invoke<{
							transactionId?: string;
						}>('app_postTransaction', {
							transaction: encodedTransaction,
						});

						this.logger.debug('Sent Report misbehavior transaction', result.transactionId);
					}
				} catch (error) {
					this.logger.error(error);
				}
			}
		});
	}

	private async _createPoMTransaction(
		contradictingBlock: chain.BlockHeader,
		decodedBlockHeader: chain.BlockHeader,
	): Promise<string> {
		// ModuleID:5 (DPoS), AssetID:3 (PoMAsset)
		const pomAssetInfo = this.apiClient.schemas.commands.find(
			({ moduleID, commandID }) => moduleID === 5 && commandID === 3,
		);

		if (!pomAssetInfo) {
			throw new Error('PoM asset schema is not registered in the application.');
		}

		// Assume passphrase is checked before calling this function
		const passphrase = this._state.passphrase as string;

		const authAccount = await this.apiClient.invoke<{ nonce: string }>('auth_getAuthAccount', {
			address: getAddressFromPassphrase(passphrase).toString('hex'),
		});

		const pomTransactionParams = {
			header1: decodedBlockHeader,
			header2: contradictingBlock,
		};

		const { networkIdentifier } = await this.apiClient.invoke<{ networkIdentifier: string }>(
			'app_getNodeInfo',
		);

		const encodedParams = codec.encode(pomAssetInfo.schema, pomTransactionParams);

		const tx = new Transaction({
			moduleID: pomAssetInfo.moduleID,
			commandID: pomAssetInfo.commandID,
			nonce: BigInt(authAccount.nonce),
			senderPublicKey:
				this._state.publicKey ?? getAddressAndPublicKeyFromPassphrase(passphrase).publicKey,
			fee: BigInt(this.config.fee), // TODO: The static fee should be replaced by fee estimation calculation
			params: encodedParams,
			signatures: [],
		});

		(tx.signatures as Buffer[]).push(
			signData(
				TAG_TRANSACTION,
				Buffer.from(networkIdentifier, 'hex'),
				tx.getSigningBytes(),
				passphrase,
			),
		);

		return tx.getBytes().toString('hex');
	}
}
