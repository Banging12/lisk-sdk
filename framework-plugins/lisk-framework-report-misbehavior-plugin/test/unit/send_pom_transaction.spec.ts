/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { BaseChannel, GenesisConfig, testing, chain, ApplicationConfigForPlugin } from 'lisk-sdk';
import { when } from 'jest-when';

import { ReportMisbehaviorPlugin } from '../../src';
import { configSchema } from '../../src/schemas';

const appConfigForPlugin: ApplicationConfigForPlugin = {
	rootPath: '~/.lisk',
	label: 'my-app',
	logger: {
		consoleLogLevel: 'info',
		fileLogLevel: 'none',
		logFileName: 'plugin-MisbehaviourPlugin.log',
	},
	rpc: {
		modes: ['ipc'],
		ws: {
			port: 8080,
			host: '127.0.0.1',
			path: '/ws',
		},
		http: {
			port: 8000,
			host: '127.0.0.1',
		},
	},
	generation: {
		force: false,
		waitThreshold: 2,
		generators: [],
		modules: {},
	},
	network: {
		seedPeers: [],
		port: 5000,
	},
	transactionPool: {
		maxTransactions: 4096,
		maxTransactionsPerAccount: 64,
		transactionExpiryTime: 3 * 60 * 60 * 1000,
		minEntranceFeePriority: '0',
		minReplacementFeeDifference: '10',
	},
	version: '',
	networkVersion: '',
	genesis: {} as GenesisConfig,
};

const validPluginOptions = {
	...configSchema.default,
	encryptedPassphrase:
		'salt=683425ca06c9ff88a5ab292bb5066dc5&cipherText=4ce151&iv=bfaeef79a466e370e210f3c6&tag=e84bf097b1ec5ae428dd7ed3b4cce522&version=1',
	dataPath: '/my/app',
};

describe('Send PoM transaction', () => {
	let reportMisbehaviorPlugin: ReportMisbehaviorPlugin;
	const defaultNetworkIdentifier =
		'93d00fe5be70d90e7ae247936a2e7d83b50809c79b73fa14285f02c842348b3e';
	const channelMock = {
		registerToBus: jest.fn(),
		once: jest.fn(),
		publish: jest.fn(),
		subscribe: jest.fn(),
		isValidEventName: jest.fn(),
		isValidActionName: jest.fn(),
		invoke: jest.fn(),
		eventsList: [],
		actionsList: [],
		actions: {},
		moduleName: '',
		options: {},
	} as any;
	const header1 = chain.BlockHeader.fromBytes(
		Buffer.from(
			'080010fe86b28b06180022002a003220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8553a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8554220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855480050005a20f8da7f49e92286b0129fd75a9208eed942ef1d79df93c42c9b87e8b6bb9fc84f6206080012001a006a00',
			'hex',
		),
	);
	const header2 = chain.BlockHeader.fromBytes(
		Buffer.from(
			'080010bb87b28b06180022002a003220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8553a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8554220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855480050005a20d800954794e0882c2419fe4736c2a191e6515859a7a894043ba5c911da6b72e76206080012001a006a00',
			'hex',
		),
	);

	beforeEach(async () => {
		reportMisbehaviorPlugin = new ReportMisbehaviorPlugin();
		await reportMisbehaviorPlugin.init({
			config: validPluginOptions,
			channel: (channelMock as unknown) as BaseChannel,
			appConfig: appConfigForPlugin,
			logger: testing.mocks.loggerMock,
		});
		(reportMisbehaviorPlugin as any)._channel = channelMock;
		(reportMisbehaviorPlugin as any)._options = { fee: '100000000' };
		reportMisbehaviorPlugin['logger'] = {
			error: jest.fn(),
		} as any;
		jest.spyOn(reportMisbehaviorPlugin['apiClient'], 'schemas', 'get').mockReturnValue({
			block: chain.blockSchema,
			blockHeader: chain.blockHeaderSchema,
			transaction: chain.transactionSchema,
			commands: [
				{
					moduleID: 5,
					moduleName: 'dpos',
					commandID: 3,
					commandName: 'reportDelegateMisbehavior',
					schema: {
						$id: 'lisk/dpos/pom',
						type: 'object',
						required: ['header1', 'header2'],
						properties: {
							header1: {
								...chain.blockHeaderSchema,
								$id: 'block-header1',
								fieldNumber: 1,
							},
							header2: {
								...chain.blockHeaderSchema,
								$id: 'block-header2',
								fieldNumber: 2,
							},
						},
					},
				},
			],
		});
		(reportMisbehaviorPlugin as any)._state = {
			passphrase: testing.fixtures.defaultFaucetAccount.passphrase,
			publicKey: testing.fixtures.defaultFaucetAccount.publicKey,
		};

		when(jest.spyOn(reportMisbehaviorPlugin['apiClient'], 'invoke'))
			.calledWith('auth_getAuthAccount', expect.anything())
			.mockResolvedValue({ nonce: '0' } as never)
			.calledWith('app_getNodeInfo')
			.mockResolvedValue({ networkIdentifier: defaultNetworkIdentifier } as never);
	});

	it('should throw error when pom transaction params schema is not found', async () => {
		jest.spyOn(reportMisbehaviorPlugin['apiClient'], 'schemas', 'get').mockReturnValue({
			block: chain.blockSchema,
			blockHeader: chain.blockHeaderSchema,
			transaction: chain.transactionSchema,
			commands: [],
		});
		await expect(
			(reportMisbehaviorPlugin as any)._createPoMTransaction(header1, header2),
		).rejects.toThrow('PoM params schema is not registered in the application.');
	});

	it('should create pom transaction for given block headers', async () => {
		const pomTransaction = await (reportMisbehaviorPlugin as any)._createPoMTransaction(
			header1,
			header2,
		);
		expect(pomTransaction).toMatchSnapshot();
	});
});
