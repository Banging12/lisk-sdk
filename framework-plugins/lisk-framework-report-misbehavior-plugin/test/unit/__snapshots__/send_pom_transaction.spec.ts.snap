// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send PoM transaction should create pom transaction for given block headers 1`] = `"0805100318002080c2d72f2a200fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a32ce020aa401080010bb87b28b06180022002a003220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8553a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8554220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855480050005a20d800954794e0882c2419fe4736c2a191e6515859a7a894043ba5c911da6b72e76206080012001a006a0012a401080010fe86b28b06180022002a003220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8553a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8554220e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855480050005a20f8da7f49e92286b0129fd75a9208eed942ef1d79df93c42c9b87e8b6bb9fc84f6206080012001a006a003a40614ae816c7ced020d00a4260c1980e2353f3ccdaea4e4054c903e0fa2cc6ba1ce462821cfe38958a9f96c92631f0629bf244151c2bb7b1bd7b96c2395174890b"`;
